torchtune/training/_activation_offloading.py

```diff
@@ -0,0 +1,308 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from warnings import warn
+
+import psutil
+import torch
+from torch.autograd.graph import saved_tensors_hooks
+
+
+class OffloadActivations(saved_tensors_hooks):
+    """Context manager under which activation tensors created in the forward pass will be offloaded.
+
+    Enable the memory efficiency technique of activation offloading, where activations bigger than
+    min_offload_size bytes will be offloaded to CPU in the forward and brought back in the backward.
+    This is in contrast to maintaining the activation on GPU VRAM throughout the program.
+
+    This manager contains the option of using one additional CUDA stream to handle the communication
+    between CUDA and CPU, which is intended to overlap with the default computation stream to improve
+    runtime. We designed synchronization with a few heuristics for optimizing the tradeoff between
+    runtime vs memory usage.
+
+    Args:
+        use_pin_memory (bool): Whether or not the offloaded Tensor will be placed in pinned
+            memory on the CPU. Pinned memory allows the Tensor to be moved back onto GPU more quickly
+            but is a limited resource. Default: True.
+
+        use_streams (bool): Whether or not to use streams for performance optimization where
+            the communications get overlapped with the computation. Default: True.
```

Comment:
we should probably add some safe guards here for torch 2.4. Is that true @ebsmothers, or people cant use nightlies with 2.4?

Reply:
100% I think it'd have to be 2.5 too, since the change needed is in nightlies and not even 2.4.

Reply:
Oh yeah let's definitely add a version check then. What will the behavior be if someone tries to run on 2.4? Will it fully error out? And is it only a concern when `use_streams=True`?

Reply:
my understanding is that use_streams=False works in 2.4, so i dont see why it should error. A warning plus defaulting to use_streams=False sounds reasonable to me

Reply:
Yea, use_streams=False will work with 2.4. It's only with use_streams=True where we should error. Is there already another feature gated by torch version that I should copy?

And yes, I will default to False for use_streams for this PR